nupic/research/frameworks/pytorch/imagenet/mixins/vary_batch_size.py

```diff
@@ -0,0 +1,104 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2020, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from copy import deepcopy
+
+
+class VaryBatchSize(object):
+    """
+    This mixin enables loading the training data with varying batch-sizes from
+    epoch to epoch.
+
+    :param config:
+        - batch_sizes: list of batch sizes; the last one will be used for the remainder
+                       of training once all have been exhausted, going in order
+        - batch_size: Can be given as a trivial case where the batch size doesn't vary;
+                      can only be used when `batch_sizes=None`
+    """
+
+    def setup_experiment(self, config):
+
+        # Validate batch sizes.
+        batch_sizes = config.get("batch_sizes", None)
+        batch_size = config.get("batch_size", None)
+
+        # Assert exclusive or; only one can't be None.
+        assert (batch_size is not None) != (batch_sizes is not None), (
+            "Must specify one of 'batch_size' or 'batch_sizes', not both.")
+
+        # Set the default list of batch sizes where the won't vary epoch to epoch.
+        if batch_size is not None:
+            batch_sizes = [batch_size]
+        # Ensure batch_sizes is a list.
+        if batch_sizes is not None:
+            assert isinstance(batch_sizes, list), "Must specify list of batch sizes"
+
+        self.batch_sizes = batch_sizes
+        super().setup_experiment(config)
```

Comment:
Why not calling super at the beginning of the method?

Reply:
`setup_experiment` calls `create_train_dataloader`, and wanted to be sure the args were validated before this happens.